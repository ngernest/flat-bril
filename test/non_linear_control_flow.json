{
  "functions": [
    {
      "args": [
        {
          "name": "b0",
          "type": "bool"
        },
        {
          "name": "b1",
          "type": "bool"
        }
      ],
      "instrs": [
        {
          "labels": [
            "start"
          ],
          "op": "jmp"
        },
        {
          "label": "end"
        },
        {
          "args": [
            "x_0_2"
          ],
          "op": "print"
        },
        {
          "args": [
            "x_1_2"
          ],
          "op": "print"
        },
        {
          "op": "ret"
        },
        {
          "label": "l_1_3"
        },
        {
          "labels": [
            "end"
          ],
          "op": "jmp"
        },
        {
          "label": "l_1_2"
        },
        {
          "dest": "x_1_2",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "labels": [
            "l_1_3"
          ],
          "op": "jmp"
        },
        {
          "label": "l_1_1"
        },
        {
          "dest": "x_1_1",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "labels": [
            "l_1_3"
          ],
          "op": "jmp"
        },
        {
          "label": "l_0_3"
        },
        {
          "args": [
            "b1"
          ],
          "labels": [
            "l_1_1",
            "l_1_2"
          ],
          "op": "br"
        },
        {
          "label": "l_0_2"
        },
        {
          "dest": "x_0_2",
          "op": "const",
          "type": "int",
          "value": 2
        },
        {
          "labels": [
            "l_0_3"
          ],
          "op": "jmp"
        },
        {
          "label": "l_0_1"
        },
        {
          "dest": "x_0_1",
          "op": "const",
          "type": "int",
          "value": 3
        },
        {
          "labels": [
            "l_0_3"
          ],
          "op": "jmp"
        },
        {
          "label": "start"
        },
        {
          "args": [
            "b0"
          ],
          "labels": [
            "l_0_1",
            "l_0_2"
          ],
          "op": "br"
        }
      ],
      "name": "main"
    }
  ]
}
