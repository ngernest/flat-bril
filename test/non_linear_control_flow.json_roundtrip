{
  "functions": [
    {
      "args": [
        {
          "name": "b0",
          "type": "bool"
        },
        {
          "name": "b1",
          "type": "bool"
        }
      ],
      "instrs": [
        {
          "args": [],
          "funcs": [],
          "labels": [
            "start"
          ],
          "op": "jmp"
        },
        {
          "label": "end"
        },
        {
          "args": [
            "x_0_2"
          ],
          "funcs": [],
          "labels": [],
          "op": "print"
        },
        {
          "args": [
            "x_1_2"
          ],
          "funcs": [],
          "labels": [],
          "op": "print"
        },
        {
          "args": [],
          "funcs": [],
          "labels": [],
          "op": "ret"
        },
        {
          "label": "l_1_3"
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "end"
          ],
          "op": "jmp"
        },
        {
          "label": "l_1_2"
        },
        {
          "dest": "x_1_2",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "l_1_3"
          ],
          "op": "jmp"
        },
        {
          "label": "l_1_1"
        },
        {
          "dest": "x_1_1",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "l_1_3"
          ],
          "op": "jmp"
        },
        {
          "label": "l_0_3"
        },
        {
          "args": [
            "b1"
          ],
          "funcs": [],
          "labels": [
            "l_1_1",
            "l_1_2"
          ],
          "op": "br"
        },
        {
          "label": "l_0_2"
        },
        {
          "dest": "x_0_2",
          "op": "const",
          "type": "int",
          "value": 2
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "l_0_3"
          ],
          "op": "jmp"
        },
        {
          "label": "l_0_1"
        },
        {
          "dest": "x_0_1",
          "op": "const",
          "type": "int",
          "value": 3
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "l_0_3"
          ],
          "op": "jmp"
        },
        {
          "label": "start"
        },
        {
          "args": [
            "b0"
          ],
          "funcs": [],
          "labels": [
            "l_0_1",
            "l_0_2"
          ],
          "op": "br"
        }
      ],
      "name": "main"
    }
  ]
}
