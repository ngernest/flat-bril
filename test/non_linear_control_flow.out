@main
	op: jmp  		labels: ["start"]
.end
	op: print		args: ["x_0_2"]	
	op: print		args: ["x_1_2"]	
	op: ret  	
.l_1_3
	op: jmp  		labels: ["end"]
.l_1_2
	op: const		dest: x_1_2		type: int		value: 0	
	op: jmp  		labels: ["l_1_3"]
.l_1_1
	op: const		dest: x_1_1		type: int		value: 1	
	op: jmp  		labels: ["l_1_3"]
.l_0_3
	op: br   		args: ["b1"]		labels: ["l_1_1", "l_1_2"]
.l_0_2
	op: const		dest: x_0_2		type: int		value: 2	
	op: jmp  		labels: ["l_0_3"]
.l_0_1
	op: const		dest: x_0_1		type: int		value: 3	
	op: jmp  		labels: ["l_0_3"]
.start
	op: br   		args: ["b0"]		labels: ["l_0_1", "l_0_2"]
